metricbeat/mb/mb.go

```diff
@@ -94,6 +104,44 @@ func (m *BaseModule) UnpackConfig(to interface{}) error {
 	return m.rawConfig.Unpack(to)
 }
 
+// ReConfigure re-configures the module with the given raw configuration
+func (m *BaseModule) ReConfigure(config *common.Config, register *Register) error {
+	currName := m.name
+	currConfig := m.config
+	currRawConfig := m.rawConfig
+
+	var chkConfig struct {
+		Module string `config:"module"`
+	}
+	if err := config.Unpack(&chkConfig); err != nil {
+		return errors.Wrap(err, "error parsing new module configuration")
+	}
+
+	// Don't allow module name change
+	if chkConfig.Module != "" && chkConfig.Module != currName {
+		return fmt.Errorf("cannot change module name from %v to %v", currName, chkConfig.Module)
+	}
+
+	if err := config.SetString("module", -1, currName); err != nil {
+		return errors.Wrap(err, "unable to set existing module name in new configuration")
+	}
+
+	if err := config.Unpack(&m.config); err != nil {
+		return errors.Wrap(err, "error parsing new module configuration")
+	}
+	m.rawConfig = config
+
+	// Calling initMetricSets ensures proper validation checks
+	if _, err := initMetricSets(register, m); err != nil {
+		m.name = currName
+		m.config = currConfig
+		m.rawConfig = currRawConfig
```

Comment:
If we are concerned about modifying the original base module, maybe we can make this method to return a copy of the base module with the modifications, and keep the original base module the same. So the interface of this method could be something like this:
```
func (m *BaseModule) WithConfig(config *common.Config) (*BaseModule, error)
```
And in the same line, `elastic.ConfigureModule` would also return a copy of the module, this time as an `mb.Module`, not intended to be reconfigured anymore:
```
func ConfigureModule(base *mb.BaseModule, xpackEnabledMetricsets []string) (mb.Module, error)
```
This returned module could then be returned by `NewModule()` methods.

Reply:
I think this is much better. Will change, thanks!

Reply:
Implemented in 525c46260.